hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheRecordStore.java

```diff
@@ -1445,41 +1443,67 @@ public void removeAll(Set<Data> keys, int completionId) {
     }
 
     @Override
-    public CacheRecord merge(CacheMergeTypes mergingEntry, SplitBrainMergePolicy<Data, CacheMergeTypes> mergePolicy) {
-        final long now = Clock.currentTimeMillis();
-        final long start = isStatisticsEnabled() ? System.nanoTime() : 0;
+    public boolean merge(CacheMergeTypes mergingEntry, SplitBrainMergePolicy<Object, CacheMergeTypes> mergePolicy) {
+        long now = Clock.currentTimeMillis();
+        long start = isStatisticsEnabled() ? System.nanoTime() : 0;
 
         injectDependencies(mergingEntry);
         injectDependencies(mergePolicy);
 
-        boolean merged = false;
-        Data key = mergingEntry.getKey();
+        Data dataKey = mergingEntry.getKey();
         long expiryTime = mergingEntry.getExpirationTime();
-        R record = records.get(key);
-        boolean isExpired = processExpiredEntry(key, record, now);
+        R record = records.get(dataKey);
+        Data dataValue = record == null ? null : toHeapData(record.getValue());
+
+        CacheMergeTypes existingEntry = createMergingEntry(nodeEngine.getSerializationService(), dataKey, dataValue, record);
```

Comment:
I like how the method is now easier to read with the early return. I see the previous version had the optimisation that it first checked if the record was expired before copying the value on-heap and creating the merging entry. Can we do the same here? Maybe it is also a matter of correctness, since the merge policy will be under the impression the value is still non-expired (and may return it) but later we check and see that it is, in fact, expired.

Reply:
Good one, I fixed that and also removed the `return null` parts from the `MergingValueFactory` in this PR.